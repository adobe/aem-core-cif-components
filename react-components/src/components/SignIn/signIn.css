.root {
    display: grid;
    gap: 1.5rem;
    justify-items: stretch;
    padding: 1rem 1.5rem;
}

.forgotPassword {
    display: block;
    text-decoration: underline;
    color: rgb(var(--venia-error));
    font-size: 0.8rem;
}

.form {
    display: grid;
    row-gap: 1rem;
}

.formTitle {
    padding: 1rem 0 0.5rem;
}

.modal {
    visibility: hidden;
    height: 100%;
    width: 100%;
    background-color: rgb(var(--venia-grey));
    text-align: center;
    position: absolute;
    bottom: 0;
}

.modal_active {
    visibility: visible;
    composes: modal;
    opacity: 0.9;
}

.createAccountButton,
.forgotPasswordButton,
.signInButton {
    align-items: center;
    display: flex;
    text-align: center;
}

.createAccountButton,
.signInButton {
    justify-content: center;
}

.createAccountButton button,
.signInButton button {
    min-width: 10rem;
}

.forgotPasswordButton {
    justify-content: flex-start;
}

.signInButton {
    margin-top: 1rem;
}

.signInError {
    background-color: rgb(var(--venia-grey));
    border-radius: 4px;
    color: rgb(var(--venia-error));
    font-size: 0.875rem;
    line-height: 1.25rem;
    padding: 1rem;
}

.signInError:empty {
    display: none;
}
