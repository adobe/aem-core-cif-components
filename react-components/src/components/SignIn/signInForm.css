.form {
    display: grid;
    row-gap: 1rem;
}

.signInError {
    background-color: rgb(var(--venia-grey));
    border-radius: 4px;
    color: rgb(var(--venia-error));
    font-size: 0.875rem;
    line-height: 1.25rem;
    padding: 1rem;
}

.signInError:empty {
    display: none;
}

.signInButton {
    align-items: center;
    display: flex;
    justify-content: center;
    text-align: center;
    margin-top: 1rem;
}
