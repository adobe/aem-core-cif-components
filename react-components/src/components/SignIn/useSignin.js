/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { useState } from 'react';
import { useUserContext } from '../../context/UserContext';
import { useMutation } from '@apollo/client';
import { useCartState } from '../Minicart/cartContext';
import { useAwaitQuery, useCookieValue } from '../../utils/hooks';
import { mergeCarts } from '../../actions/cart';

import MUTATION_MERGE_CARTS from '../../queries/mutation_merge_carts.graphql';
import QUERY_CUSTOMER_CART from '../../queries/query_customer_cart.graphql';
import MUTATION_GENERATE_TOKEN from '../../queries/mutation_generate_token.graphql';
import QUERY_CART_DETAILS from '../../queries/query_cart_details.graphql';

export const useSignin = props => {
    const { showMyAccount } = props;
    const [{ cartId }, cartDispatch] = useCartState();
    const [userState, { setToken, getUserDetails, setCustomerCart, setError }] = useUserContext();
    const [inProgress, setInProgress] = useState(false);

    const [, setCartCookie] = useCookieValue('cif.cart');

    const [mergeCartsMutation] = useMutation(MUTATION_MERGE_CARTS);
    const fetchCustomerCart = useAwaitQuery(QUERY_CUSTOMER_CART);
    const cartDetailsQuery = useAwaitQuery(QUERY_CART_DETAILS);
    const [generateCustomerToken] = useMutation(MUTATION_GENERATE_TOKEN);

    let errorMessage = '';
    if (userState.signInError && userState.signInError.length > 0) {
        errorMessage = userState.signInError;
    }

    const handleSubmit = async ({ email, password }) => {
        setInProgress(true);
        try {
            // 1. generate the customer token
            const { data } = await generateCustomerToken({ variables: { email, password } });
            const token = data.generateCustomerToken.token;
            setToken(token);

            // 2. get the user details
            await getUserDetails();

            const { data: customerCartData } = await fetchCustomerCart({
                fetchPolicy: 'network-only'
            });
            const customerCartId = customerCartData.customerCart.id;

            // 3. merge the shopping cart if necessary
            let mergedCartId;

            if (cartId) {
                mergedCartId = await mergeCarts({
                    mergeCartsMutation,
                    cartDetailsQuery,
                    cartId,
                    customerCartId,
                    dispatch: cartDispatch
                });
            } else {
                mergedCartId = customerCartId;
            }
            //4. set the cart id in the cookie
            setCartCookie(mergedCartId);
            setCustomerCart(mergedCartId);

            //5. show my account view in account dropdown or navigation side panel after sign in
            showMyAccount();
        } catch (e) {
            setError(e);
            setInProgress(false);
        }
    };

    return {
        inProgress,
        handleSubmit,
        errorMessage,
        isSignedIn: userState.isSignedIn
    };
};
