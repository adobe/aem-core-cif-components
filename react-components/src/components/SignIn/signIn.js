/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { Form } from 'informed';
import { func } from 'prop-types';
import { useTranslation } from 'react-i18next';

import { isRequired } from '../../utils/formValidators';
import Button from '../Button';
import Field from '../Field';
import TextInput from '../TextInput';

import classes from './signIn.css';
import { useSignin } from './useSignin';
import LoadingIndicator from '../LoadingIndicator';

const SignIn = props => {
    const { showMyAccount, showForgotPassword, showCreateAccount } = props;
    const { errorMessage, handleSubmit, inProgress } = useSignin({ showMyAccount });
    const [t] = useTranslation('account');

    if (inProgress) {
        return <LoadingIndicator>{t('account:signing-in', 'Signing In')}</LoadingIndicator>;
    }

    return (
        <div className={classes.root}>
            <Form onSubmit={handleSubmit} className={classes.form}>
                <div className={classes.formTitle}>{t('account:sign-in-form-title', 'Sign-in to Your Account')}</div>
                <Field label={t('account:email', 'Email address')} required={true}>
                    <TextInput autoComplete="email" field="email" validate={isRequired} aria-label="email" />
                </Field>
                <Field label={t('account:password', 'Password')} required={true}>
                    <TextInput
                        autoComplete="current-password"
                        field="password"
                        type="password"
                        validate={isRequired}
                        aria-label="password"
                    />
                </Field>
                <div className={classes.forgotPasswordButton}>
                    <Button priority="low" type="button" onClick={showForgotPassword}>
                        {t('account:forgot-password', 'Forgot Password?')}
                    </Button>
                </div>
                <div className={classes.signInError}>{errorMessage}</div>
                <div className={classes.signInButton}>
                    <Button priority="high" type="submit" aria-label="submit">
                        {t('account:sign-in', 'Sign In')}
                    </Button>
                </div>
                <div className={classes.createAccountButton}>
                    <Button priority="normal" type="button" onClick={showCreateAccount}>
                        {t('account:create-an-account', 'Create an Account')}
                    </Button>
                </div>
            </Form>
        </div>
    );
};

SignIn.propTypes = {
    showMyAccount: func.isRequired,
    showForgotPassword: func.isRequired,
    showCreateAccount: func.isRequired
};

export default SignIn;
