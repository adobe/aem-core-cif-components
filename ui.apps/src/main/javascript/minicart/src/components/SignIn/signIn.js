/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { Form } from 'informed';
import { func } from 'prop-types';

import { isRequired } from '../../utils/formValidators';
import Button from '../Button';
import Field from '../Field';
import TextInput from '../TextInput';

import classes from './signIn.css';
import { useSignin } from './useSignin';

const SignIn = props => {
    const { showMyAccount } = props;
    const { errorMessage, isSignedIn, handleSubmit } = useSignin();

    if (isSignedIn) {
        showMyAccount();
    }

    return (
        <div className={classes.root}>
            <Form onSubmit={handleSubmit} className={classes.form}>
                <Field label="Email" required={true}>
                    <TextInput autoComplete="email" field="email" validate={isRequired} aria-label="email" />
                </Field>
                <Field label="Password" required={true}>
                    <TextInput
                        autoComplete="current-password"
                        field="password"
                        type="password"
                        validate={isRequired}
                        aria-label="password"
                    />
                </Field>
                <div className={classes.signInError}>{errorMessage}</div>
                <div className={classes.signInButton}>
                    <Button priority="high" type="submit">
                        {'Sign In'}
                    </Button>
                </div>
            </Form>
        </div>
    );
};

SignIn.propTypes = {
    showMyAccount: func.isRequired
};

export default SignIn;
