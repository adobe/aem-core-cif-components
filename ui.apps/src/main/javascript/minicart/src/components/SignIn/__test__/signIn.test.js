/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React, { useState } from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';

import UserContextProvider from '../../../context/UserContext';

import MUTATION_GENERATE_TOKEN from '../../../queries/mutation_generate_token.graphql';
import QUERY_CUSTOMER_DETAILS from '../../../queries/query_customer_details.graphql';

import SignIn from '../signIn';

const mocks = [
    {
        request: {
            query: MUTATION_GENERATE_TOKEN,
            variables: {
                email: 'chuck@example.com',
                password: 'norris'
            }
        },
        result: {
            data: {
                generateCustomerToken: {
                    token: 'token-123'
                }
            }
        }
    },
    {
        request: {
            query: QUERY_CUSTOMER_DETAILS
        },
        result: {
            data: {
                customer: {
                    email: 'test@example.com',
                    firstname: 'John',
                    lastname: 'Doe'
                }
            }
        }
    }
];

describe('<SignIn>', () => {
    beforeEach(() => {
        Object.defineProperty(window.document, 'cookie', {
            writable: true,
            value: ''
        });
    });

    it('renders the component', () => {
        const { asFragment } = render(
            <MockedProvider>
                <UserContextProvider>
                    <SignIn showMyAccount={jest.fn()} showCreateAccount={jest.fn()} showForgotPassword={jest.fn()} />
                </UserContextProvider>
            </MockedProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
    it('switch the view then the "Sign In" is successful', async () => {
        // To simulate an almost real use case of the sign in component we create a wrapper around it
        // which displays a "success" message when the user is signed in
        const SignInWrapper = () => {
            const [signedIn, setSignedIn] = useState(false);

            const showMyAccount = () => {
                setSignedIn(true);
            };
            let content;
            if (signedIn) {
                content = <div data-testid="success">Done</div>;
            } else {
                content = (
                    <SignIn
                        showMyAccount={showMyAccount}
                        showCreateAccount={jest.fn()}
                        showForgotPassword={jest.fn()}
                    />
                );
            }

            return <div>{content}</div>;
        };

        const { getByTestId, getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <UserContextProvider>
                    <SignInWrapper />
                </UserContextProvider>
            </MockedProvider>
        );

        fireEvent.change(getByLabelText(/email/i), { target: { value: 'chuck@example.com' } });
        fireEvent.change(getByLabelText(/password/i), { target: { value: 'norris' } });
        fireEvent.click(getByLabelText('submit'));

        const result = await waitForElement(() => getByTestId('success'));

        expect(result.textContent).not.toBeUndefined();
    });

    it('shows an error when the sign in is not successful', async () => {
        const mocks = [
            {
                request: {
                    query: MUTATION_GENERATE_TOKEN,
                    variables: {
                        email: 'chuck@example.com',
                        password: 'wrongpassword'
                    }
                },
                result: {
                    data: {
                        generateCustomerToken: null
                    },
                    errors: [
                        {
                            message: 'Error',
                            category: 'graphql-authentication'
                        }
                    ]
                }
            }
        ];

        const { getByText, getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <UserContextProvider>
                    <SignIn showMyAccount={jest.fn()} showForgotPassword={jest.fn()} showCreateAccount={jest.fn()} />
                </UserContextProvider>
            </MockedProvider>
        );

        fireEvent.change(getByLabelText(/email/i), { target: { value: 'chuck@example.com' } });
        fireEvent.change(getByLabelText(/password/i), { target: { value: 'wrongpassword' } });
        fireEvent.click(getByLabelText('submit'));

        const result = await waitForElement(() => getByText('Error'));

        expect(result).not.toBeUndefined();
    });
});
