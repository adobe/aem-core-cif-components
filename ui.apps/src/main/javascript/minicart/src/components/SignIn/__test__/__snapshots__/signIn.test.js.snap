// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn> renders the component 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <form
      class="form"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          <span
            class="requiredSymbol"
          />
          Email
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              aria-label="email"
              autocomplete="email"
              class="input"
              name="email"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
      <div
        class="root"
      >
        <span
          class="label"
        >
          <span
            class="requiredSymbol"
          />
          Password
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              aria-label="password"
              autocomplete="current-password"
              class="input"
              name="password"
              type="password"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
      <div
        class="signInError"
      />
      <div
        class="signInButton"
      >
        <button
          class="root_highPriority"
          type="submit"
        >
          <span
            class="content"
          >
            Sign In
          </span>
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
